crates/clvm-utils/src/tree_hash.rs

```diff
@@ -293,319 +373,437 @@ pub fn tree_hash_cached(a: &Allocator, node: NodePtr, cache: &mut TreeCache) ->
+            hash = tree_hash_pair(tree_hash_atom(b"\x01"), hash);
+            cache.insert(list, &hash, 0);
+
+            println!("{i}");
+            if i < 65533 {
+                let (h, _c) = cache.get(list).expect("expected cached");
+                assert_eq!(h, &hash);
+            } else {
+                assert_eq!(cache.get(list), None);
+            }
+        }
+        assert_eq!(cache.get(list), None);
     }
-}
 
-#[test]
-fn test_tree_cache_get() {
-    let mut allocator = Allocator::new();
-    let mut cache = TreeCache::default();
+    #[test]
+    fn test_tree_cache_should_memoize() {
+        let mut allocator = Allocator::new();
+        let mut cache = TreeCache::default();
 
-    let a = allocator.nil();
-    let b = allocator.one();
-    let c = allocator.new_pair(a, b).expect("new_pair");
+        let a = allocator.nil();
+        let b = allocator.one();
+        let c = allocator.new_pair(a, b).expect("new_pair");
 
-    assert_eq!(cache.get(a), None);
-    assert_eq!(cache.get(b), None);
-    assert_eq!(cache.get(c), None);
+        assert!(!cache.should_memoize(a));
+        assert!(!cache.should_memoize(b));
+        assert!(!cache.should_memoize(c));
 
-    // We don't cache atoms
-    cache.insert(a, &tree_hash(&allocator, a));
-    assert_eq!(cache.get(a), None);
+        // we need to visit a node at least twice for it to be considered a
+        // candidate for memoization
+        assert!(cache.visit(c));
+        assert!(!cache.should_memoize(c));
+        assert!(!cache.visit(c));
 
-    cache.insert(b, &tree_hash(&allocator, b));
-    assert_eq!(cache.get(b), None);
+        assert!(cache.should_memoize(c));
+    }
 
-    // but pair is OK
-    cache.insert(c, &tree_hash(&allocator, c));
-    assert_eq!(cache.get(c), Some(&tree_hash(&allocator, c)));
-}
+    #[test]
+    fn test_tree_hash_costed_equivalence_no_repeats() {
+        let mut a = Allocator::new();
 
-#[test]
-fn test_tree_cache_size_limit() {
-    let mut allocator = Allocator::new();
-    let mut cache = TreeCache::default();
+        // Build a nested tree:
+        // ((a . b) . ((x . y) . (z . w)))
+        let a_atom = a.new_atom(b"a").unwrap();
+        let b_atom = a.new_atom(b"b").unwrap();
+        let x_atom = a.new_atom(b"x").unwrap();
+        let y_atom = a.new_atom(b"y").unwrap();
+        let z_atom = a.new_atom(b"z").unwrap();
+        let w_atom = a.new_atom(b"w").unwrap();
+
+        let ab_pair = a.new_pair(a_atom, b_atom).unwrap();
+        let xy_pair = a.new_pair(x_atom, y_atom).unwrap();
+        let zw_pair = a.new_pair(z_atom, w_atom).unwrap();
+        let right_pair = a.new_pair(xy_pair, zw_pair).unwrap();
+        let root = a.new_pair(ab_pair, right_pair).unwrap();
+
+        let mut cost_left_baseline = 1_000_000;
+        let mut cost_left_cached = 1_000_000;
+
+        // baseline: costed but no caching
+        let baseline = tree_hash_costed(&a, root, &mut cost_left_baseline).unwrap();
```

Comment:
I think the baseline could just be when you run it with an empty cache. If you think the baseline calculation should happen with a separate function, I think this function should move into this test, and be a test-only function.